### ISSUE TITLE
Missing rrules on x86 Julia for matrix distributions

### ISSUE THREAD

**Issue by penelopeysm** [compintell/Mooncake.jl#289]
I'll preface this by saying that these excerpts are being run on x64 Linux but with an x86 Julia binary, so I'm not sure how relevant / reproducible they really are, but it's causing [some Turing test failures upstream](https://github.com/TuringLang/Turing.jl/actions/runs/11293905299/job/31413153549?pr=2341), so...

The different distributions in the MWE below result in slightly different types in the error traceback. For Wishart it's `Tuple{Val{(:dpotrf_, Symbol("libblastrampoline.so.5"))},...`, for MatrixBeta it's `:dgetrf`, and for the MvNormals it's `:dtrtrs`.

```julia
using Distributions: Wishart, MatrixBeta, MvNormal
using DynamicPPL: @model, LogDensityFunction
using LogDensityProblems: logdensity_and_gradient
using LogDensityProblemsAD: ADgradient
import Mooncake

d = Wishart(7, [1 0.5; 0.5 1])

# These also fail
# d = MatrixBeta(3, 4, 4)
# d = MvNormal([0. ;0.; 0.], [1. 0. 0.; 0. 1. 0.; 0. 0. 1.])
# d = MvNormal([0.; 0.], [1. 0.; 0. 1.])

@model mat() = x ~ d
ℓ = ADgradient(:Mooncake, LogDensityFunction(mat()))
logdensity_and_gradient(ℓ, vec(rand(d)))
```

<details><summary>Full traceback (for Wishart)</summary>

```
ERROR: No rrule!! available for foreigncall with primal argument types Tuple{Val{(:dpotrf_, Symbol("libblastrampoline.so.5"))}, Val{Nothing}, Tuple{Val{Ref{UInt8}}, Val{Ref{Int32}}, Val{Ptr{Float64}}, Val{Ref{Int32}}, Val{Ptr{Int32}}, Val{Int32}}, Val{0}, Val{:ccall}, Ptr{UInt8}, Ptr{Int32}, Ptr{Float64}, Ptr{Int32}, Ptr{Int32}, Int32, Int32, Base.RefValue{Int32}, Base.RefValue{Int32}, Matrix{Float64}, Base.RefValue{Int32}, Base.RefValue{UInt8}}. This problem has most likely arisen because there is a ccall somewhere in the function you are trying to differentiate, for which an rrule!! has not been explicitly written.You have three options: write an rrule!! for this foreigncall, write an rrule!! for a Julia function that calls this foreigncall, or re-write your code to avoid this foreigncall entirely. If you believe that this error has arisen for some other reason than the above, or the above does not help you to workaround this problem, please open an issue.
Stacktrace:
  [1] rrule!!(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/rrules/foreigncall.jl:12
  [2] potrf!
    @ /opt/hostedtoolcache/julia/1.10.5/x86/share/julia/stdlib/v1.10/LinearAlgebra/src/lapack.jl:3217 [inlined]
  [3] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
  [4] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
  [5] DerivedRule
    @ ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734 [inlined]
  [6] (::Mooncake.LazyDerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:1509
  [7] #cholesky!#140
    @ /opt/hostedtoolcache/julia/1.10.5/x86/share/julia/stdlib/v1.10/LinearAlgebra/src/cholesky.jl:268 [inlined]
  [8] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
  [9] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
 [10] DerivedRule
    @ ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734 [inlined]
 [11] (::Mooncake.LazyDerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:1509
 [12] #cholesky!#141
    @ /opt/hostedtoolcache/julia/1.10.5/x86/share/julia/stdlib/v1.10/LinearAlgebra/src/cholesky.jl:296 [inlined]
 [13] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
 [14] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
 [15] DerivedRule
    @ ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734 [inlined]
 [16] (::Mooncake.LazyDerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:1509
 [17] nonsingular_wishart_logkernel
    @ ~/.julia/packages/Distributions/uuqsE/src/matrix/wishart.jl:189 [inlined]
 [18] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
 [19] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
 [20] DerivedRule
    @ ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734 [inlined]
 [21] (::Mooncake.LazyDerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:1509
 [22] invlink_with_logpdf
    @ ~/.julia/packages/DynamicPPL/ooLj8/src/abstract_varinfo.jl:782 [inlined]
 [23] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
 [24] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
 [25] DerivedRule
    @ ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734 [inlined]
 [26] (::Mooncake.LazyDerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:1509
 [27] tilde_assume!!
    @ ~/.julia/packages/DynamicPPL/ooLj8/src/context_implementations.jl:144 [inlined]
 [28] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
 [29] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
 [30] (::Mooncake.DerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734
 [31] (::Mooncake.DynamicDerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:1447
 [32] (::Mooncake.RRuleZeroWrapper{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:247
 [33] mat
    @ ./REPL[26]:1 [inlined]
 [34] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
 [35] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
 [36] (::Mooncake.DerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734
 [37] (::Mooncake.LazyDerivedRule{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ Mooncake ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:1509
 [38] RRuleZeroWrapper
    @ ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:247 [inlined]
 [39] logdensity
    @ ~/.julia/packages/DynamicPPL/ooLj8/src/logdensityfunction.jl:138 [inlined]
 [40] (::Tuple{…})(none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…}, none::Mooncake.CoDual{…})
    @ Base.Experimental ./<missing>:0
 [41] (::MistyClosures.MistyClosure{…})(::Mooncake.CoDual{…}, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
    @ MistyClosures ~/.julia/packages/MistyClosures/rzVHC/src/MistyClosures.jl:15
 [42] DerivedRule
    @ ~/.julia/packages/Mooncake/GMFJI/src/interpreter/s2s_reverse_mode_ad.jl:734 [inlined]
 [43] logdensity_and_gradient(∇l::MooncakeLogDensityProblemsADExt.MooncakeGradientLogDensity{…}, x::Vector{…})
    @ MooncakeLogDensityProblemsADExt ~/.julia/packages/Mooncake/GMFJI/ext/MooncakeLogDensityProblemsADExt.jl:48
 [44] top-level scope
    @ REPL[28]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

</details>

<details><summary>versioninfo()</summary>

(This was run on a github actions runner)

```
julia> versioninfo()
Julia Version 1.10.5
Commit 6f3fdf7b362 (2024-08-27 14:19 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (i686-linux-gnu)
  CPU: 4 × AMD EPYC 7763 64-Core Processor
  WORD_SIZE: 32
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, znver3)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
```
</details>

**Comment by willtebbutt** [comment]
This looks quite odd, because I'm quite sure that I've added rules for all of these things. There must be something slightly off with the rule implementations -- I'll take a look once the 1.11 stuff is resolved.

**Comment by willtebbutt** [comment]
I believe this to be resolved on 0.4.16 . The problem was just that the name associated to the BLAS / LAPACK function being `ccall`ed varies depending on whether you're running a 64bit vs 32bit system. I had never run Mooncake on 32 bit, so I had missed this.

I'll leave it to you to close this issue if the Turing.jl problems have been resolved!

**Comment by penelopeysm** [comment]
Indeed, it's [fixed in Turing CI](https://github.com/TuringLang/Turing.jl/actions/runs/11449590718/job/31855473155?pr=2341) :) Thanks Will!
